### PR TITLE
refactor: make all udf function impls public

### PR DESCRIPTION
**Pull request by universalmind303**
## Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/9900

## What changes are included in this PR?

this just makes all of the structs implementing `ScalarUDFImpl` publicly accessible

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Makes sense to me -- thank you @universalmind303 

I wonder if you want to try and add some sort of test to ensure these are all pub somehow? If not, I predict that as we port over additional functions some will get missed

File: datafusion/functions-array/src/array_has.rs [hunk 1]
```diff
@@ -54,11 +54,17 @@ make_udf_function!(ArrayHasAny,
 );
 
 #[derive(Debug)]
-pub(super) struct ArrayHas {
```
Comment (jayzhan211) [inline datafusion/functions-array/src/array_has.rs:57]: It seems that all the arrays function are public, do you plan to public the rest of them too?

Comment (universalmind303) [reply datafusion/functions-array/src/array_has.rs:57]: Yes, this should include all of functions from `functions` and `functions-array` crates at the time of authoring.

**Comment by alamb** [comment]
Thanks @universalmind303. BTW @viirya  is discussing adding a way to instantiate only some of the UDFs here https://github.com/apache/arrow-datafusion/issues/9892

**Comment by alamb** [comment]
I merged this branch up locally to main and verified clippy and tests passed. 🚀

**Pull request merged**